File: src/sexpr.jl
"For writing Autumn programs, prior to having an Autumn parser"
module SExpr
# using Rematch
using MLStyle
using SExpressions
using ..AExpressions

export parseau, @au_str, parseautumn


fg(s) = s
fg(s::Cons) = array(s)
"Convert an `SExpression` into nested Array{Any}"
array(s::SExpression) = [map(fg, s)...]

@inline rest(sexpr::SExpressions.Cons) = sexpr.cdr
@inline rest2(sexpr::SExpressions.Cons) = rest(rest(sexpr))
"""Parse string `saexpr` into AExpr

```julia

prog = \"\"\"
(program
  (external (:: x Int))
  (:: y Float64)
  (group Thing (:: position Int) (:: alpha Bool))
  (= y 1.2)
  (= 
    (-> (x y)
        (let (z (+ x y))
              (* z y)))
)
\"\"\"

"""
function parseautumn(sexprstring::AbstractString)
  io = IOBuffer(sexprstring)
  nocomments_sexprstring = ""
  for line in eachline(io)
    nocomments_sexprstring *= (split(line, ";")[1] * "\n")
  end

  parseau(array(SExpressions.Parser.parse(nocomments_sexprstring)))
  # parseau(array(SExpressions.Parser.parse(sexprstring)))
end

"Parse SExpression into Autumn Expressions"
function parseau(sexpr::AbstractArray)
  res = MLStyle.@match sexpr begin
    [:program, lines...]              => AExpr(:program, map(parseau, lines)...)
    [:if, c, :then, t, :else, e]      => AExpr(:if, parseau(c), parseau(t), parseau(e))
    [:initnext, i, n]                 => AExpr(:initnext, parseau(i), parseau(n))
    [:(=), x::Symbol, y]              => AExpr(:assign, x, parseau(y))
    [:(:), v::Symbol, τ]              => AExpr(:typedecl, v, parsetypeau(τ))
    [:external, tdef]                 => AExpr(:external, parseau(tdef))
    [:let, vars]                      => AExpr(:let, map(parseau, vars)...)
    [:case, name, cases...]           => AExpr(:case, name, map(parseau, cases)...)
    [:(=>), type, value]              => AExpr(:casevalue, parseau(type), parseau(value))
    [:type, :alias, var, val]         => AExpr(:typealias, var, parsealias(val))
    [:fn, params, body]               => AExpr(:fn, AExpr(:list, params...), parseau(body))
    [:(-->), var, val]                => AExpr(:lambda, parseau(var), parseau(val))
    [:list, vars...]                  => AExpr(:list, map(parseau, vars)...)
    [:.., var, field]                 => AExpr(:field, parseau(var), parseau(field))
    [:on, args...]                    => AExpr(:on, map(parseau, args)...)
    [:object, args...]                => AExpr(:object, map(parseau, args)...)  # note that this isn't super clear wrt naming
    [:include, path]                  => AExpr(:include, path)
    [:import, name]                   => AExpr(:import, name)
    [:structure, name::Symbol, args...] => AExpr(:structure, name, map(parseau, args)...) # Julia literally uses Symbol for variable names
    [:module, name::Symbol, lines...] => AExpr(:module, name, map(parseau, lines)...)
    [f, xs...]                        => AExpr(:call, parseau(f), map(parseau, xs)...)
    [vars...]                         => AExpr(:list, map(parseau, vars)...)
  end
end

# function parseletvars(list::Array{})
#   result = []
#   i = 1
#   while i < length(list)
#     push!(result, AExpr(:assign, parseau(list[i]), parseau(list[i+1])))
#     i += 2
#   end
#   push!(result, parseau(list[length(list)]))
#   result
# end

function parsealias(expr)
  AExpr(:typealiasargs, map(parseau, expr)...)
end

#(: map (-> (-> a b) (List a) (List b)))
function parsetypeau(sexpr::AbstractArray)
  MLStyle.@match sexpr begin
    [τ, tvs...] && if (istypesymbol(τ) && all(istypevarsymbol.(tvs)))   end => AExpr(:paramtype, τ, tvs...)
    [:->, τs...]                                                            => AExpr(:functiontype, map(parsetypeau, τs)...)
    [args...]                                                               => [args...]
  end
end

parseau(list::Array{Int, 1}) = list[1]
parsetypeau(s::Symbol) = s
parseau(s::Symbol) = s
parseau(s::Union{Number, String}) = s

"""
Macro for parsing autumn
au\"\"\"
(program
  (= x 3)
  (let (x 3) (+ x 3))
)
\"\"\"
"""
macro au_str(x::String)
  QuoteNode(parseautumn(x))
end

end

File: src/js/sexpr.js
var parse = require('s-expression');

// Construct Autumn Expression (AExpr) from program string
export function parseau(program_str) {
  sexpr = parse(program_str);
  aexpr = s_to_a(sexpr); 
  return aexpr;
}

// Convert S-Expression (SExpr) to Autumn Expression (AExpr)
function s_to_a(sexpr) {
  if (Array.isArray(sexpr)) { 
    if (sexpr[0] == "program") {
      return {"head" : "", "args" : sexpr.slice(1).map(line => s_to_a(line))};
    } else if (sexpr[0] == "if") {
      return {"head" : "if", "args" : [s_to_a(sexpr[1]), s_to_a(sexpr[3]), s_to_a(sexpr[5])]};
    } else if (sexpr[0] == "initnext") {
      return {"head" : "initnext", "args" : [s_to_a(sexpr[1]), s_to_a(sexpr[2])]};
    } else if (sexpr[0] == "=") {
      return {"head" : "assign", "args" : [sexpr[1], s_to_a(sexpr[2])]};
    } else if (sexpr[0] == ":") {
      return {"head" : "typedecl", "args" : [sexpr[1], s_to_a(sexpr[2])]};
    } else if (sexpr[0] == "let") {
      return {"head" : "let", "args" : sexpr[1].map(line => s_to_a(line))};
    } else if (sexpr[0] == "fn") {
      return {"head" : "fn", "args" : [{"head" : "list", "args" : sexpr[1]}, s_to_a(sexpr[2])]};
    } else if (sexpr[0] == "-->") {
      return {"head" : "lambda", "args" : [s_to_a(sexpr[1]), s_to_a(sexpr[2])]};
    } else if (sexpr[0] == "..") {
      return {"head" : "field", "args" : [s_to_a(sexpr[1]), s_to_a(sexpr[2])]};
    } else if (sexpr[0] == "on") {
      return {"head" : "on", "args" : [s_to_a(sexpr[1]), s_to_a(sexpr[2])]};
    } else if (sexpr[0] == "object") {
      return {"head" : "object", "args" : [s_to_a(sexpr[1]), s_to_a(sexpr[2])]};
    } else if (sexpr.length > 1) {
      return {"head" : "call", "args" : [s_to_a(sexpr[0]), sexpr.slice(1).map(s => s_to_a(s))]};
    } else {
      return {"head" : "list", "args" : sexpr.map(elt => s_to_a(elt))};
    } 
  } else if (!isNaN(sexpr)) {
    return parseInt(sexpr);
  } else if (typeof(sexpr) == "string") {
    return sexpr;
  } else {
    return {"str" : sexpr.valueOf()};
  }
}